packages/cli/src/databases/utils/migrationHelpers.ts

```diff
@@ -72,6 +84,42 @@ export const wrapMigration = (migration: Migration) => {
 
 	const { up, down } = migration.prototype;
 	Object.assign(migration.prototype, {
+		async beforeTransaction(this: typeof migration.prototype, queryRunner: QueryRunner) {
+			if (
+				this.pruneAndVacuum &&
+				dbType === 'sqlite' &&
+				process.env.ENABLE_MIGRATIONS_PRUNING === 'true'
+			) {
+				const dbFileSize = getSqliteDbFileSize();
+				if (dbFileSize < DESIRED_DATABASE_FILE_SIZE) {
+					return;
+				}
+				const counting = (await queryRunner.query(
+					`select count(id) as rows from "${tablePrefix}execution_entity";`,
+				)) as Array<{ rows: number }>;
+
+				const averageExecutionSize = dbFileSize / counting[0].rows;
+				const nubmerOfExecutionsToKeep = Math.floor(
+					DESIRED_DATABASE_FILE_SIZE / averageExecutionSize,
+				);
+
+				const query = `
+					SELECT id FROM "${tablePrefix}execution_entity"
+					ORDER BY id DESC limit ${nubmerOfExecutionsToKeep}, 1;
+				`;
+
+				const idToKeep = (await queryRunner.query(query)) as Array<{ id: number }>;
+
+				const removalQuery = `
+					DELETE FROM "${tablePrefix}execution_entity"
+					WHERE id < ${idToKeep[0].id} and status IN ("success");
```

Comment:
We're not loading all iDs, the query contains a limit 1 to return the single ID where the cut is and then we perform a delete from that point onwards.

I did this to avoid having a subquery as it can be very slow on some database systems.